### PR TITLE
optimize AggregateGradientsImpl (async pipeline) for non-GDR & GPU sc…

### PR DESCRIPTION
**Pull request by juncgu**
This patch optimizes AggregateGradientsImpl for non-GDR & GPU mode. In our experiment, this can cut almost 50% of aggregation time when training a model with 60 million parameters (using DataParallelSGD, gradientBits=32, MPI=openMPI).

For non-GDR + GPU scenario, the original aggregator uses async GPU/CPU copy and nonblocking (async) allreduce. However, the original pipeline is not efficient with those batched async operations. Because nonblocking MPI calls may not work until MPI Wait() is called. 
original-pipeline

We changed the pipeline using async GPU/CPU copy and blocking allreduce for overlapping data copy and transmission. 
Our new pipeline for aggregator is:
new-pipeline

This commit has been tested under non-GDR + GPU.  
We may also need a CopyGPUToCPUSync() in GPUDataTransferer.

Thank you,
Juncheng

### CODE REVIEW

File: Source/SGDLib/SimpleDistGradAggregator.h [hunk 1]
```diff
@@ -326,36 +305,100 @@ class SimpleDistGradAggregator : public IDistGradAggregator<ElemType>
         if (!m_mpi->IsMainNode())
             m_mpi->Isend(headerCPU, headerCPU->Size(), MPI_CHAR, m_mpi->MainNodeRank(), numGradMatrices, &sendHeaderRequest) || MpiFail("MPI_Isend");
 
+
+        // New aggregation pipeline for non-GDR
         // Perform async allreduce on the gradient data
```
Comment (junjieqian) [inline Source/SGDLib/SimpleDistGradAggregator.h]: >async allreduce on the  [](start = 19, length = 23)

This would be "sync allreduce".

File: Source/SGDLib/SimpleDistGradAggregator.h [hunk 2]
```diff
@@ -326,36 +305,100 @@ class SimpleDistGradAggregator : public IDistGradAggregator<ElemType>
...
+        size_t cpuToGpuIndex = 0;
+        size_t allReduceIndex = 0;
+        // non-GDR && GPU
+        if ((m_mpi->UseGpuGdr() == 0) && (deviceId != CPUDEVICE))
         {
-            size_t allReduceIndex = 0;
+            Matrix<ElemType>* gpuCopyBuffer = m_aggregationBuffer.get();
+
             ElemType* reductionBuffer;
+            size_t j;
```
Comment (junjieqian) [inline Source/SGDLib/SimpleDistGradAggregator.h]: >j [](start = 19, length = 1)

`j` is not initialized but used in `m_mpi->AllReduce`. Also can you rename `j` to some meaningful name?

File: Source/SGDLib/SimpleDistGradAggregator.h (hunk 1 above)
Comment (juncgu) [inline Source/SGDLib/SimpleDistGradAggregator.h]: Yeah. My comment is not accurate. For non-GDR & GPU, we use sync allreduce; but for CPU, I make no modification, still async allreduce. I will update the words.

File: Source/SGDLib/SimpleDistGradAggregator.h (hunk 2 above)
Comment (juncgu) [inline Source/SGDLib/SimpleDistGradAggregator.h]: j is the current gradientIndex we need to aggregate through AllReduce.  I will change j to "currentGradientIndex", and init it with the first index in m_gradientIndexToAggregate.

File: Source/SGDLib/SimpleDistGradAggregator.h [hunk 3]
```diff
@@ -326,36 +305,105 @@ class SimpleDistGradAggregator : public IDistGradAggregator<ElemType>
...
+                cpuToGpuIndex = allReduceIndex;
+                m_gpuDataTransferers[cpuToGpuIndex]->CopyCPUToGPUAsync(m_intermediateCPUBuffers[cpuToGpuIndex].get(),
+                    (currentGradientIndex == -1) ? m_aggregationBuffer->GetNumElements() : gradients[currentGradientIndex]->GetNumElements(),
+                    (currentGradientIndex == -1) ? m_aggregationBuffer->Data() : gradients[currentGradientIndex]->Data());
+                allReduceIndex ++;
+                gpuToCpuIndex++;
+                currentGradientIndex = i;
+            }
+            // If currentGradientIndex == -2, it means no gradient index from m_gradientIndexToAggregate
+            if (currentGradientIndex == -2)
```
Comment (junjieqian) [inline Source/SGDLib/SimpleDistGradAggregator.h]: This seems incorrect. `currentGradientIndex == -2`, (1) it should never happen that no gradients to aggregate (until this part); (2) it contradicts with following codes (line 379...). Can you check this part again? Or we should sync again on this.

File: Source/SGDLib/SimpleDistGradAggregator.h [hunk 4]
```diff
@@ -326,36 +305,105 @@ class SimpleDistGradAggregator : public IDistGradAggregator<ElemType>
...
-                    reductionBuffer = m_intermediateCPUBuffers[allReduceIndex].get();
+                    gpuCopyBuffer = gradients[i];
+                }
+                else
+                {
+                    // i == -1, first element is for packed gradients, which should not be with AsyncAggregation
+                    assert(m_useAsyncAggregation == false);
+                }
+                //1.first sync_g_to_c_copy
+                if (gpuToCpuIndex == 0)
```
Comment (junjieqian) [inline Source/SGDLib/SimpleDistGradAggregator.h]: Prefer to move this (copy from GPU to CPU for the first gradient) out of the loop. So the logic is cleaner that (1) async copy for the first one; (2) in the loop, wait for last GPU->CPU copy, start copy GPU->CPU for next one, do aggregation, start copy CPU->GPU; (3) out the loop, wait for CPU->GPU (you already did this).

File: Source/SGDLib/SimpleDistGradAggregator.h [hunk 5]
```diff
@@ -326,36 +305,105 @@ class SimpleDistGradAggregator : public IDistGradAggregator<ElemType>
...
+        // non-NCCL 
+        else if (!m_nccl.IsSupported())
+        {
+            ElemType* reductionBuffer;
+            for (size_t i : m_gradientIndexToAggregate)
+            {
+                allReduceRequests.push_back(MPI_Request());
+                reductionBuffer = (i == -1)? m_aggregationBuffer->Data() : gradients[i]->Data();
+                // CPU
+                if (m_mpi->UseGpuGdr() ==0)
```
Comment (junjieqian) [inline Source/SGDLib/SimpleDistGradAggregator.h]: nit: space `== 0`.

File: Source/SGDLib/SimpleDistGradAggregator.h (hunk 4 above)
Comment (juncgu) [inline Source/SGDLib/SimpleDistGradAggregator.h]: Thank you. I agree that the logic will be more clear. But about "(2) in the loop", I think the order of operations should be "start async copy GPU->CPU for next gradient, wait for last GPU->CPU copy, do aggregation, start async CPU-GPU".  "Wait" shouldn't block the next GPU->CPU copy. What's your opinion?

Comment (junjieqian) [inline Source/SGDLib/SimpleDistGradAggregator.h]: Make sense.

File: Source/SGDLib/SimpleDistGradAggregator.h (hunk 3 above)
Comment (juncgu) [inline Source/SGDLib/SimpleDistGradAggregator.h]: Sorry, it's my bug. It should be "if (currentGradientIndex != -2)" in line 366. It is intended to handle no gradients case (if possible). Fixed in commit ada7743829834cbec63d5744962c41f70189dfe7

File: Source/SGDLib/SimpleDistGradAggregator.h (hunk 5 above)
Comment (juncgu) [inline Source/SGDLib/SimpleDistGradAggregator.h]: Thank you! I will follow the format in the future. Fixed in commit ada7743829834cbec63d5744962c41f70189dfe7

File: Source/SGDLib/SimpleDistGradAggregator.h [hunk 6]
```diff
@@ -312,105 +312,99 @@ class SimpleDistGradAggregator : public IDistGradAggregator<ElemType>
...
-                //2. starting from 2nd loop : 2.1 async copy next 
-                m_gpuDataTransferers[gpuToCpuIndex]->CopyGPUToCPUAsync(gpuCopyBuffer->Data(), gpuCopyBuffer->GetNumElements(), m_intermediateCPUBuffers[gpuToCpuIndex].get());
+                // First sync_g_to_c_copy
+                // TODO: we need a CopyGPUToCPUSync
+                cudaMemcpy(m_intermediateCPUBuffers[gpuToCpuIndex].get(), gpuCopyBuffer->Data(), gpuCopyBuffer->GetNumElements() * sizeof(ElemType), cudaMemcpyDeviceToHost);
+                gpuToCpuIndex++;
 
-                //2.2 sync Allreduce
-                if (allReduceIndex != 0)
+                for (i = 1; i <= numGradientIndex; i ++)
```
Comment (junjieqian) [inline Source/SGDLib/SimpleDistGradAggregator.h]: where is `i` defined?

Comment (juncgu) [inline Source/SGDLib/SimpleDistGradAggregator.h]: Sorry, my mistake. Fixed in commit fe913854e27712e83fa2ddfbe6c937345c28ea80

**Comment by junjieqian** [comment]
I created branch `juncgu/optimizingAgg` and it passed the Jenkins tests. @KeDengMS Can you take a quick look?

**Comment by KeDengMS** [comment]
@junjieqian, the change looks fine to me. Please merge with latest master and submit through Jenkins after 2.2 release.

**Comment by juncgu** [comment]
Thank you so much, @junjieqian and @KeDengMS.

**Comment by KeDengMS** [comment]
Tried branch  juncgu/optimizingAgg in Jenkins and a test keeps failing:

21:28:27 Speech/DNN/Parallel1BitQuantizationWithAutoMBScaling: MPI Rank 1: 09/18/2017 21:28:09:  Epoch[ 1 of 3]-Minibatch[ 661- 670, 52.34%]: CrossEntropyWithSoftmax = 2.82608420 * 640; EvalClassificationError = 0.70937500 * 640; time = 0.0550s; samplesPerSecond = 11640.7
21:28:27 Speech/DNN/Parallel1BitQuantizationWithAutoMBScaling: MPI Rank 1: 09/18/2017 21:28:09:  Epoch[ 1 of 3]-Minibatch[ 671- 680, 53.12%]: CrossEntropyWithSoftmax = 2.86517133 * 640; EvalClassificationError = 0.67343750 * 640; time = 0.0541s; samplesPerSecond = 11839.2
21:28:27 [FAILED]: Testcase Per-minibatch training results must match for each MPI Rank
21:28:27 Baseline: MPI Rank 1: 05/18/2017 11:20:04:  Epoch[ 1 of 3]-Minibatch[ 671- 680, 53.12%]: CrossEntropyWithSoftmax = 2.86512788 * 640; EvalClassificationError = 0.67187500 * 640; time = 0.0712s; samplesPerSecond = 8990.6
21:28:27 Failed pattern: EvalClassificationError = {{float,tolerance=0.1%}}

@juncgu, can you try the test under tests/EndToEndTests/Speech/dnn/Parallel1BitQuantizationWithAutoMBScaling?

**Comment by juncgu** [comment]
Sure.

**Comment by junjieqian** [comment]
merged. Thanks for contribution!

**Comment by juncgu** [comment]
Thank you!
